host_source.go

```diff
@@ -271,106 +365,279 @@ func checkSystemLocal(control *controlConn) (bool, error) {
 		}
 
-		if r.matchFilter(host) {
-			hosts = append(hosts, host)
+		// If it's not a valid peer
+		if !r.IsValidPeer(host) {
+			Logger.Printf("Found invalid peer '%+v' "+
+				"Likely due to a gossip or snitch issue, this host will be ignored", host)
+			continue
 		}
+		hosts = append(hosts, host)
+	}
+	if it.err != nil {
+		return nil, errors.Wrap(it.err, "GetClusterPeerInfo()")
+	}
+	return hosts, nil
+}
+
+// Return true if the host is a valid peer
+func (r *ringDescriber) IsValidPeer(host *HostInfo) bool {
+	return !(len(host.RPCAddress()) == 0 ||
+		host.hostId == "" ||
+		host.dataCenter == "" ||
+		host.rack == "" ||
+		len(host.tokens) == 0)
+}
+
+// Return a list of hosts the cluster knows about
+func (r *ringDescriber) GetHosts() ([]*HostInfo, string, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// Update the localHost info with data from the connected host
+	localHost, err := r.GetLocalHostInfo()
+	if err != nil {
+		return r.prevHosts, r.prevPartitioner, err
+	}
+
+	// Update our list of hosts by querying the cluster
+	hosts, err := r.GetClusterPeerInfo()
+	if err != nil {
+		return r.prevHosts, r.prevPartitioner, err
 	}
 
-	if err = rows.Err(); err != nil {
-		return nil, "", err
+	hosts = append(hosts, localHost)
+
+	// Filter the hosts if filter is provided
+	var filteredHosts []*HostInfo
+	if r.session.cfg.HostFilter != nil {
+		for _, host := range hosts {
+			if r.session.cfg.HostFilter.Accept(host) {
+				filteredHosts = append(filteredHosts, host)
+			}
+		}
+	} else {
+		filteredHosts = hosts
 	}
 
-	r.prevHosts = hosts
-	r.prevPartitioner = partitioner
+	r.prevHosts = filteredHosts
+	r.prevPartitioner = localHost.partitioner
+	r.localHost = localHost
 
-	return hosts, partitioner, nil
+	return filteredHosts, localHost.partitioner, nil
 }
 
-func (r *ringDescriber) matchFilter(host *HostInfo) bool {
-	if r.dcFilter != "" && r.dcFilter != host.DataCenter() {
-		return false
+// Given an ip/port return HostInfo for the specified ip/port
+func (r *ringDescriber) GetHostInfo(ip net.IP, port int) (*HostInfo, error) {
+	var host *HostInfo
+	var err error
+
+	// TODO(thrawn01): Is IgnorePeerAddr still useful now that we have DisableInitialHostLookup?
+	// TODO(thrawn01): should we also check for DisableInitialHostLookup and return if true?
+
+	// Ignore the port and connect address and use the address/port we already have
+	if r.session.control == nil || r.session.cfg.IgnorePeerAddr {
+		return &HostInfo{connectAddress: ip, port: port}, nil
+	}
+
+	// Attempt to get the host info for our control connection
+	controlHost := r.session.control.GetHostInfo()
+	if controlHost == nil {
+		return nil, errors.New("invalid control connection")
+	}
+
+	// If we are asking about the same node our control connection has a connection too
+	if controlHost.ConnectAddress().Equal(ip) {
+		host, err = r.GetLocalHostInfo()
+	} else {
+		host, err = r.GetPeerHostInfo(ip, port)
+	}
+	// Always respect the user provided or discovered address
```

Comment:
I evaluated this statement again and found that the java driver appears to use the address provided by the `NEW_NODE` event for peer table lookup only. I will change the code to mirror this and improve the comment to better explain why we prefer the user provided addresses in the case of the control connection.